### PR TITLE
Add flexible normalization method

### PR DESCRIPTION
**Pull request by rguliev**
The previous pull does not contain my changes :)
For some reason, I thought it would better to cancel my changes, then merge current master, and then add my changes. Unfortunately, you accepted my previous pull request when I had canceled my changes. Sorry, for my weird actions.
This is new pull request with my changes. One commit - one file is added.

### CODE REVIEW

**Comment by cbeleites** [comment]
Hi Rustam,
well, I didn't accept the pull request, but I merged master into the new_methods branch and apparently at a point when you had canceled your changes, so the pull request has 0 diff and I guess that's why it vanished.

**Comment by cbeleites** [comment]
Rustam, I'm sorry, I was caught basically without internet (i.e. only superslow mobile internet working). I'll contact you next week when I'm available for a skype/slack call.

**Comment by rguliev** [comment]
Hi Claudia,
I'm not an active slack user, let's use skype then. You can find me there by email.

**Comment by rguliev** [comment]
I still haven't received a message or call. Any problem in finding me?

**Comment by cbeleites** [comment]
Rustam, did you get my email? => I didn't get any answer from you, so now I'm trying to make contact this way. I'd be available tomorrow, say, 1pm CEST or after 5pm CEST or on Thursday (hard to predict times, though). OR, you could suggest a few possibilites that work for you, this week or next week.

**Comment by rguliev** [comment]
Oh.. I didn't receive any email from you. I even checked spam. I thought you were too busy) Ok, how about 6pm CEST tomorrow?

**Comment by cbeleites** [comment]
Summary of skype conversation:

- functions like `normalize` will be useful particularly to make `hyperSpec` working smoothly with pipe operators / `magrittr`
- In order to have the normalization options (area / minmax / vector / ...) not restricted to the bunch of functions Rustam thought about, it would be good to design the functionality as 
  - [ ] `normalize` being a gateway similar to `apply` to which some function is handed over that does the actual calculation.  
  this gateway function needs to have 2 parameters for spectra, similar to  `apply.to` and `fit.to` in `spc.poly`: 
    - `apply.to` being the spectra that should be normlized
    - `fit.to`being used to calculate the normalization coefficients (and possibly offsets), default would be `apply.to`
 This allows to use the same function e.g. either to area normalize on the whole spectrum, or on a particular peak/band.
  - [ ] instead of the `switch` statement, the "workhorse" function handed over should be called on `apply.to` and `fit.to`

**Comment by GegznaV** [comment]
My suggestion about the names of functions: please, consider package-specific function names, e.g., `spc_normalize()`, `spc.normalize()` or similar., as e.g., `stringr::str_replace()`, to avoid possible name space conflicts.

**Comment by cbeleites** [comment]
@GegznaV: that's a good point. `spc_` prefix is one possibility. OTOH, we could also keep the workhorse function names internal to hyperSpec's namespace (no export). People can still look at them with the `:::` operator, and then the only function that needs to be exported would be `normalize` or `spc.normalize` (to keep consistent with the baseline function names). The main advantage that a user could supply their own normalization function (which of course would need to have a different name from the internal ones) and use that one would still be available.

**Comment by rguliev** [comment]
Hi,

1) I would choose `spc.normalize` because `hyperSpec` already has similar functions. So, it's better to stick with one naming convention.

2) I know we have discussed it. But after our conversation, I kept thinking about the `apply`-like style. I still agree that it better to make it flexible, i.e. to not be restricted. On the other hand, one question keeps bugging me: "How often is this flexibility needed?". I mean, as far as I know, types area/vector/peak/01 cover about 90% of cases. So, I was thinking about what we can do here and a few options came to my mind:

- To create a function `spc.normalize(spc, FUN, ...)` as we discussed and add wrappers like `spc.normalize.area`.
- To create one function `spc.normalize` which will decide according to the second argument: if it's a function, then function will be applied; if it's a string with type (like 'area'), then corresponding normalization will be applied
- To create one function `spc.normalize` with all parameters, i.e. like, for example, `spc.normalize(spc, type='area', FUN=NULL, fit.to=NULL, apply.to=NULL)` and then decide according to parameters provided.

What would you say?

P.S. Sorry again, I know we already discussed it, but I still want to help the others who are not that good in programming by making `normalize` a clear and explicit function. If you still don't like the idea, it's ok, I will make a wrapper in my repo.

**Comment by rguliev** [comment]
I think your example proves my point that a group of few normalization types covers most cases. So, if several people came to the same independently shouldn't we consider it as something useful?:)

By the way, I didn't know that wavelength range can be sent to a function in this way. Thank you!

**Comment by cbeleites** [comment]
Well, let me make the point that I'd include some more functions into the standard normalization techniques. 

- SNV (although I usually don't like it) and MSC / EMSC immediately come to my mind. 
- People do research/develop new preprocessing methods, including normalization. Particular interest IMHO goes into methods that handle full spectra and automatically detect suitable spectral ranges.  
I think we should accommodate also this use case of hyperSpec.

Two points that I see in favor of the `apply`-like style:

- I consider this the typical R approach. Sure, it is unusual in the sense that many other programming languages do not have (or use as extensively as R) the concept of handing over functions as parameters of functions. But we're in R, where this is a typical/expected/important (and very powerful) concept. 
- The `apply`-style implementation covers the `"argument"`-style case almost on the fly: 
   - `do.call` takes both functions or function names:  

         f <- function (fun, ...){
           do.call (fun, list (...))
         }
         
         f ("mean", 1:3)
         f (mean, 1:3)

  - or we can find out the mode of `fun` and if necessary look for the function:

         g <- function (fun, ...){
           if (! mode (fun) == "function")
             fun <- get (fun)
           fun (...)
         }
         
         g ("mean", 1:3)
         g (mean, 1:3)

**Comment by cbeleites** [comment]
Some more thoughts: 

- In R, function `center` with argument `scale` does basically the transpose of our normalization wrapper.  
[ ] we should mention this in the documentation.
- `center` attaches the computed `center` and `scale` vectors. This is important as those parameters are needed e.g. in cross validation, where the `center` and `scale` vectors computed from training data are applied to the test data.
- some normalization methods such as (E)MSC with default parameters (i.e. calculate reference spectrum on the fly as average spectrum) have the same behaviour. => we need to give access to the vector with normalization factors.

- The wrapper for normalization would just as well work for baseline-correction methods. And methods like MSC/EMSC/SNV that do both baseline correction and normalization.  
This is neat: it would allow us to just provide `spc_baseline <- spc_normalize` that covers both preprocessing steps (plus baseline correction methods), and maybe even the name `spc_preproc` would be more appropriate. However, this sounds to me as if it would make sense to put this into its own package (with dependency `hyperSpec`): thus it would be easy to provide all kinds of normalization and baseline functions without too much trouble about namespace conflicts.  
I forgot to mention in the call: I've met Kristian (maintainer of `baseline` package) some years ago, and he was fine with the idea to make the 2 packages work together more smoothly. 

- Only few normalization functions are needed if the normalization can rely on the fact that its input spectra are already baseline corrected. If not, the variety explodes as e.g. the area under a peak/band depends on the baseline fit in that spectral range.  
I'm all in favor of providing simple defaults (e.g. linear fit), but the flexibility to do as needed should remain. 

I'd like to hear your thoughts on this.

**Comment by GegznaV** [comment]
Yesterday I started to write (but finally didn't send) a comment about how I imagine what normalization function should do.

- I see **normalization of a spectrum** as a procedure that uses information of only one spectrum, calculates a constant from that spectrum and divides all points of the spectrum by that number. (Here is a question: do we all have the same meaning for words "spectra normalization"?)

- Because of this, I imagine that offset correction and baseline correction should be implemented as separate functions. Thus  "0-1 normalization" function, in my opinion, should be consisted of two separate functions. (Or we can have tree functions as well). 
 
- Next, as scaling depends on information from a group of spectra, it should be implemented as a separate function too. It can be associated with an appropriate `predict` method.

- To sum up, I am in favor of having many simple functions that do only **one** thing and can be joined in a chain using, e.g., `%>%`, versus one "super" function that does everything. So, my question is: do we want several simple functions or one "all-in-one" function?

**Comment by rguliev** [comment]
Hi,

> Two points that I see in favor of the apply-like style:

My suggestion is not to refuse the `apply`-like style, it is to combine both. I even suggested 3 ways of how it can be done. Maybe there is a better option of doing it. Anyway, the point is that it's possible.

> I'd like to hear your thoughts on this.

GegznaV was ahead of me:) In general, I agree with him. 

As for normalization, I would also separate "normalization" using only one spectrum (like area) and using all spectra together (like SNV). And as for me, "normalize" should cover only former one and have separate functions for the latter, i.e. `spc_snv`, `spc_emsc`. Because the outputs are a bit different.
 
As for "all-in-one", I would create another package, maybe something like `caret` extension for spectra analysis, which will have a fuction similar to `caret::preProcess` handling preprocessing in general. But implicitly it will use `hyperSpec`. Of course, such package is a quite big task and will take some time.

**Comment by cbeleites** [comment]
First of all, to remove some confusion: my center/scaling point was simply about mentioning this as a side note, I do not propose to mix in any such functionality. 

So, to summarize:

- baseline correction (including offset correction) removes additive artifacts/effects only. We use this term for methods that work on each spectrum independently of all other spectra. 
- normalization removes multiplicative artifacts/effects only. Again, we use this term for methods that work on each spectrum independently of all other spectra.  
  - It will be up to the user (though maybe aided by a guideline for the preprocessing pipeline) to make sure baseline correction was done before normalization. 
  - Such a pipeline may be provided by a function `preprocess (spc, baselineFun, normalizationFun)`
- methods that remove additive as well as multiplicative effects do not get a "group name" for now (i.e. they will have a function per method/algorithm),
- the same for methods that work on multiple spectra simultaneously. 

- We can have all this basically at no cost by exposing the lower-level "workhorse" functions as well as the high-level wrappers.

@rguliev do you want to set up such a package, or shall I set up one and make you all contributors?
